[LV] Disable fold tail by masking - when induction vars used outside

niwinanto opened the pull request:
When induction variable are used outside the loop body, tail folding
by masking mis-compiles, because for users outside of the loop the
final value of the induction is computed separately from the vector
loop.

Fixes https://github.com/llvm/llvm-project/issues/76069
Fixes https://github.com/llvm/llvm-project/issues/51677

Commit:
[LV] Disable fold tail by masking - when induction vars used outside

fhahn commented on llvm/test/Transforms/LoopVectorize/no-fold-tail-by-masking-iv-external-uses.ll:
```diff
@@ -0,0 +1,85 @@
+  br label %for.body
+
+for.cond.for.cond.cleanup_crit_edge:              ; preds = %for.body
+  %conv20.lcssa = phi i32 [ %conv20, %for.body ]
```
I think the test can be simplified by just returning `%conv20.lcssa` here

fhahn commented on llvm/test/Transforms/LoopVectorize/no-fold-tail-by-masking-iv-external-uses.ll:
```diff
@@ -0,0 +1,85 @@
+  ret void
+
+for.body:                                         ; preds = %for.body.preheader, %for.body
+  %conv20 = phi i32 [ %conv, %for.body ], [ 0, %for.body.preheader ]
```
Does the issue reproduce if all uses of `%conv20` are replaced by `i_5.019`?

fhahn commented on llvm/test/Transforms/LoopVectorize/no-fold-tail-by-masking-iv-external-uses.ll:
```diff
@@ -0,0 +1,85 @@
+
+for.body:                                         ; preds = %for.body.preheader, %for.body
+  %conv20 = phi i32 [ %conv, %for.body ], [ 0, %for.body.preheader ]
+  %i_5.019 = phi i16 [ %inc, %for.body ], [ 0, %for.body.preheader ]
```
Can the phi be changed to i32, so the `sext` in the loop isn't needed?

fhahn commented on llvm/test/Transforms/LoopVectorize/no-fold-tail-by-masking-iv-external-uses.ll:
```diff
@@ -0,0 +1,85 @@
+  %conv20 = phi i32 [ %conv, %for.body ], [ 0, %for.body.preheader ]
+  %i_5.019 = phi i16 [ %inc, %for.body ], [ 0, %for.body.preheader ]
+  %arrayidx = getelementptr inbounds [17 x i8], ptr @arr_17, i32 0, i32 %conv20
+  store i8 %val, ptr %arrayidx, align 1, !tbaa !8
```
nit: remove !tbaa metadata

fhahn commented on llvm/test/Transforms/LoopVectorize/no-fold-tail-by-masking-iv-external-uses.ll:
```diff
@@ -0,0 +1,85 @@
+; CHECK-NOT:       pred.store.continue:
+;
+entry:
+  %cmp18 = icmp sgt i32 %limit, 0
```
nit: the check and branch shouldn't be needed.

fhahn commented on llvm/test/Transforms/LoopVectorize/no-fold-tail-by-masking-iv-external-uses.ll:
```diff
@@ -0,0 +1,85 @@
+;int main(void) {
+;  int arr_2[SIZE][SIZE][SIZE];
+;
+;  __builtin_memset(arr_2, 1, sizeof(arr_2));
```
Usually we don't include C/C++ source code, as the IR usually needs to stand on its own. Below are a few suggestions to further simplify the IR and make it more readable. 

It would be helpful if you could instead a brief comment explaining the issue.

fhahn commented on llvm/test/Transforms/LoopVectorize/no-fold-tail-by-masking-iv-external-uses.ll:
```diff
@@ -0,0 +1,85 @@
+
+define void @test(i32 %limit, i8 zeroext %val, ptr readonly %arr_14)   {
+; CHECK-LABEL: @test(
+; CHECK-NOT:       pred.store.if:
```
This is quite fragile; some existing tests use `CHECK-NOT: vector.body:` to check for not vectorizing.

fhahn commented on llvm/test/Transforms/LoopVectorize/no-fold-tail-by-masking-iv-external-uses.ll:
```diff
@@ -0,0 +1,85 @@
+
+
+
+!4 = !{!5, !5, i64 0}
```
nodes used by tbaa shouldn't be needed after dropping `!tbaa`

fhahn commented on llvm/test/Transforms/LoopVectorize/no-fold-tail-by-masking-iv-external-uses.ll:
```diff
@@ -0,0 +1,85 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
+; RUN: opt < %s -passes=loop-vectorize -S | FileCheck %s
```
As this is added as a target-independent test, it probably needs something like `-force-vector-width=4 -force-vector-interleave=1` to make sure the vectorizer tries to vectorize independent of the cost-model.

fhahn reviewed (commented):
Thanks for the patch!

Could you add the test as a separate PR (with a FIXME); this patch then just adjust the test and the diff shows the change in the test only.

Previously there was a patch shared here https://reviews.llvm.org/D115109 by @rickyz  (hope it's the same as on Phabricator) but the patch never got pushed through. Would be good to look at the comments and potentially pick it up

Commit:
Code Review adjustments

niwinanto commented:
> Thanks for the patch!
> 
> Could you add the test as a separate PR (with a FIXME); this patch then just adjust the test and the diff shows the change in the test only.
> 
> Previously there was a patch shared here https://reviews.llvm.org/D115109 by @rickyz (hope it's the same as on Phabricator) but the patch never got pushed through. Would be good to look at the comments and potentially pick it up

Thanks @fhahn for the reviews. Great that you mentioned the Phabricator patch, the test looks good and I copied here.  As you suggested, created new pr for the test case with default behavior(https://github.com/niwinanto/llvm-project/pull/2/commits/33ec3081832a950369ced6d1e8913ac771e689de) and then updated this pr. However, I messed with the git workflow(I think). Could you please take a look, this is what you intended.

rickyz commented:
Thank you @niwinanto for picking this up (and apologies for letting the change languish for so long despite @fhahn's helpful comments!)

fhahn commented:
> Thanks @fhahn for the reviews. Great that you mentioned the Phabricator patch, the test looks good and I copied here. As you suggested, created new pr for the test case with default behavior([niwinanto@33ec308](https://github.com/niwinanto/llvm-project/commit/33ec3081832a950369ced6d1e8913ac771e689de)) and then updated this pr. However, I messed with the git workflow(I think). Could you please take a look, this is what you intended.

Yeah that looks good, I'll add a few small additional comments. But best to create a separate PR to just add the test case showing the issue first.

niwinanto commented:
> > Thanks @fhahn for the reviews. Great that you mentioned the Phabricator patch, the test looks good and I copied here. As you suggested, created new pr for the test case with default behavior([niwinanto@33ec308](https://github.com/niwinanto/llvm-project/commit/33ec3081832a950369ced6d1e8913ac771e689de)) and then updated this pr. However, I messed with the git workflow(I think). Could you please take a look, this is what you intended.
> 
> Yeah that looks good, I'll add a few small additional comments. But best to create a separate PR to just add the test case showing the issue first.

@fhahn I am exactly trying to create a separate PR. https://github.com/niwinanto/llvm-project/pull/2. May be you can help me to figure out what I am doing wrong. I am extremely sorry, getting used to the new workflow.

As you suggested, I created a new commit with different branch and created new PR(for test as mentioned above). For some reason it contain the commit from this PR, which I tried to remove by dropping in interactive re-base and forced push.

Also, addressed feedback regarding the tests.

fhahn commented:
> As you suggested, I created a new commit with different branch and created new PR(for test as mentioned above). For some reason it contain the commit from this PR, which I tried to remove by dropping in interactive re-base and forced push.
> 
> Also, addressed feedback regarding the tests.

Looking at https://github.com/niwinanto/llvm-project/pull/2/commits, it looks like there's a single commit adding the test, so that looks good I think? Could you update the destination branch to be upstream llvm-project's `main`?

niwinanto commented:
> > As you suggested, I created a new commit with different branch and created new PR(for test as mentioned above). For some reason it contain the commit from this PR, which I tried to remove by dropping in interactive re-base and forced push.
> > Also, addressed feedback regarding the tests.
> 
> Looking at https://github.com/niwinanto/llvm-project/pull/2/commits, it looks like there's a single commit adding the test, so that looks good I think? Could you update the destination branch to be upstream llvm-project's `main`?

@fhahn  Here is the PR, https://github.com/llvm/llvm-project/pull/82329.

niwinanto committed:
Merge branch 'main' into niwinanto/tailFoldUsedIV

Commit:
Adjust test after committing early

niwinanto commented:
@fhahn Updated the PR to adjust the changes after merging the test early.

fhahn reviewed (commented):
LGTM, thanks!

I adjusted the description of the PR a bit to add a few more details.
